src/watch_state.rs

```diff
@@ -0,0 +1,246 @@
+    /// machine if necessary.
+    pub fn handle(&self, effect: Effect) -> Outcome {
+        let mut state = self.state.lock().unwrap();
+        tracing::debug!("handling effect {:?} in current state {:?}", effect, *state);
+        let outcome = match *state {
+            State::Building => match effect {
+                Effect::Exit => {
+                    Outcome::if_running(Outcome::both(Outcome::Stop, Outcome::Exit), Outcome::Exit)
+                }
+                Effect::ProcessFailed => Outcome::if_running(Outcome::Stop, Outcome::DoNothing),
+                Effect::ProcessCompleted => {
+                    *state = State::Running;
+                    self.restart_outcome(true)
+                }
+                Effect::ManifestChange => self.restart_outcome(false),
+                Effect::SourceChange => self.restart_outcome(false),
+                Effect::ArtifactChange => Outcome::DoNothing,
+                Effect::DoNothing => Outcome::if_running(Outcome::DoNothing, Outcome::Start),
+            },
+            State::Running => match effect {
+                Effect::Exit => {
+                    Outcome::if_running(Outcome::both(Outcome::Stop, Outcome::Exit), Outcome::Exit)
+                }
+                Effect::ProcessFailed => Outcome::if_running(Outcome::Stop, Outcome::DoNothing),
+                Effect::ProcessCompleted => Outcome::DoNothing,
+                Effect::ManifestChange => {
+                    if !self.skip_build {
+                        *state = State::Building;
+                    }
+                    self.restart_outcome(false)
+                }
+                Effect::SourceChange => {
+                    if !self.skip_build {
+                        *state = State::Building;
+                        return self.restart_outcome(false);
+                    }
+                    Outcome::DoNothing
+                }
+                Effect::ArtifactChange => self.restart_outcome(false),
+                Effect::DoNothing => Outcome::if_running(Outcome::DoNothing, Outcome::Start),
+            },
+        };
+        tracing::debug!("now in {:?} state with outcome {:?}", *state, outcome);
+        outcome
+    }
+
+    fn restart_outcome(&self, prevent_clear: bool) -> Outcome {
+        let should_clear = !prevent_clear && self.clear;
+        Outcome::if_running(
+            Outcome::both(
+                Outcome::both(
+                    Outcome::Stop,
+                    match should_clear {
+                        true => Outcome::Clear,
+                        false => Outcome::DoNothing,
+                    },
+                ),
+                Outcome::Start,
+            ),
+            Outcome::both(
+                match should_clear {
+                    true => Outcome::Clear,
+                    false => Outcome::DoNothing,
+                },
+                Outcome::Start,
+            ),
+        )
```

Comment:
Took me a minute to understand what was going on here. Would this work?

```suggestion
        Outcome::sequence([
            Outcome::if_running(Outcome::Stop, Outcome::DoNothing),
            match should_clear {
                true => Outcome::Clear,
                false => Outcome::DoNothing,
            },
            Outcome::Start,
        ].into_iter())
```

Reply:
Cool. I never noticed `Outcome::sequence`